### PR TITLE
perf(codegen): fuse Range#each over literal numeric range to C for-loop

### PR DESCRIPTION
**Pull request by ryanseys**
## Summary

`(lo..hi).each { |i| ... }` over a literal numeric range previously copied the receiver into a local `sp_Range` struct and looped against `_r.first` / `_r.last`. The struct copy itself is cheap, but the looping form blocks the C compiler from seeing the bounds as constants — no unrolling, no strength reduction, no bound-check elision against an obviously-finite range.

This change recognizes literal-bounded ranges in `compile_each_block`'s range arm and emits a direct C `for` loop with bounds inlined. Runtime output is identical; the speedup is purely a compiler-visibility improvement.

```ruby
sum = 0
(1..100_000).each { |i| sum = sum + i }
puts sum
# => 5000050000
```

Pre-fix generated C:
```c
sp_Range _t2 = sp_range_new(1, 100000);
for (lv_i = _t2.first; lv_i <= _t2.last; lv_i++) { lv_sum = (lv_sum + lv_i); }
```

Post-fix generated C:
```c
for (lv_i = 1; lv_i <= 100000; lv_i++) { lv_sum = (lv_sum + lv_i); }
```

## Fix

In `compile_each_block`, when the receiver is a `RangeNode` (or a parens-wrapped one) with `IntegerNode` bounds, skip the `sp_Range` allocation entirely and emit a direct `for` loop. Inclusive vs exclusive flips between `i <= hi` and `i < hi` via the existing `range_excl_end` helper. The block-param name binding is unchanged. Non-literal ranges (`(lo..hi).each` where `lo`/`hi` are runtime expressions) continue to go through the existing `sp_Range`-allocating path.

## Benchmark

`benchmark/bm_range_each.rb` iterates `(1..500_000_000).each` accumulating `sum + (i % 7)`. The `% 7` introduces a sequential data dependency the C compiler cannot collapse to a closed form.

Methodology: `hyperfine -N --warmup 10 --runs 50` on Apple M-series, with both binaries rebuilt from clean bootstrap (`rm -f spinel_codegen build/bin* build/gen*.c build/codegen.ast && make bootstrap`).

| Command | Mean [ms] | Min [ms] | Max [ms] | Relative |
|:---|---:|---:|---:|---:|
| `master` | 345.4 ± 2.6 | 341.1 | 357.2 | 1.75 ± 0.02 |
| this PR | 197.2 ± 1.6 | 193.4 | 203.4 | 1.00 |

**Result: 1.75x ± 0.02x faster, p < 0.001.** The fused for-loop lets clang vectorize the hot path; the `sp_Range` struct loads otherwise inhibit the loop transformation.

## Out of scope

- `Range#each` over non-numeric ranges (`String`, custom `<=>`-defining objects) — Spinel doesn't model these.
- Other range-consuming methods (`Range#map`, `Range#select`) — separate per-method optimizations.
- `step` / negative-step iteration (Range#each is ascending-only by spec).
- **Block-param physical renaming when the name collides with an outer local.** Spinel's `compile_each_block` (both this fusion path and the generic `sp_Range` path) emits the block param as `lv_<name>`, which aliases an outer `lv_<name>` of the same name. The same-name shadowing case (`outer_v = 42; (1..5).each { |outer_v| }; puts outer_v`) prints `6` instead of CRuby's `42` on master too — this is a pre-existing spinel-wide limitation in block-param emission, not introduced by this PR. The expanded test fixture intentionally uses distinct names (`|inner|`, etc.) to exercise scoping correctly within current capabilities.

## Test plan

- [x] `make` — builds clean
- [x] `make bootstrap` — gen2.c == gen3.c (bootstrap OK)
- [x] `make test` — full suite passes (437 / 0 fail / 0 error after the test additions)
- [x] `test/range_each_fusion.rb` covers (after this revision):
  - Inclusive / exclusive bounds
  - Negative bounds; range crossing zero (exclusive)
  - Single-element inclusive; empty exclusive (`5...5`); zero-element at the origin (`0..0`)
  - **Decreasing range** (`10..1` and `10...1`) — zero iterations, matching CRuby
  - Block-param scoping with distinct names — outer var preserved when block uses a different name
  - Variable-bounded range (forces fall-through to `sp_Range` path) — output parity
  - **Parens-wrapped range** (`((1..5)).each`) — fusion still recognizes the `RangeNode` through the `ParenthesesNode` wrapper
  - **`break`** inside the loop body — accumulator captures only pre-break values
  - **`next`** inside the loop body — even / odd skipping pattern
  - Nested ranges (2-deep) — inner loop doesn't shadow outer block param when names differ
  - **Triple-nested with mixed inclusive / exclusive** ranges
- [x] `benchmark/bm_range_each.rb` — CRuby and Spinel produce matching output (`1499999997`); hyperfine numbers reproducible via the methodology above

### CODE REVIEW

File: test/range_each_fusion.rb [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+# Block param shadows nothing — fresh local
+count = 0
+(1..100).each { |i| count = count + 1 }
+puts count
+
+# Nested ranges
+total = 0
+(1..3).each { |i| (1..3).each { |j| total = total + i * j } }
+puts total
```
[the remark this answers is not in the document]
Comment (ryanseys) [reply test/range_each_fusion.rb:104]: Thanks — three good catches. Pushed an expanded `test/range_each_fusion.rb` with coverage for two of the three you flagged plus several adjacent edge cases I think are worth pinning:

- **Decreasing range** (`(10..1)` and `(10...1)`) → 0 iterations, matching CRuby. The fused for-loop's `i <= hi` check is false on entry, so this Just Works.
- **Parens-wrapped range** (`((1..5)).each`) → fusion still applies. The recognizer unwraps `ParenthesesNode` before deciding to inline.
- **Same-name shadowing** (`outer_v = 42; (1..5).each { |outer_v| }; puts outer_v`) → I dug into this and **it's a pre-existing spinel-wide limitation**, not specific to fusion. Both this fusion path and the generic `sp_Range` path emit the block param as `lv_<name>`, which physically aliases an outer `lv_<name>` of the same name — so the same-name case prints `6` instead of CRuby's `42` on master too (verified with a variable-bounded reproducer that takes the unfused path). Fixing it requires renaming block params at emit time, which is a broader codegen change touching every block-param emission site. I've added an "Out of scope" item documenting this and adjusted the test fixture to use distinct names (`|inner|`) so the scoping test exercises something real within current capabilities.

Other additions I bundled in while reworking the fixture: zero-bound (`0..0`), range crossing zero exclusive (`-2...2`), variable-bounded fall-through path (output parity check), `break` early-exit, `next` even/odd skip, and triple-nested mixed inclusive/exclusive. All exercise both the fused path and the generic fall-through path correctly.

Pushed as `b98395d`. Test count: 437 / 0 fail / 0 error; bootstrap fixpoint unchanged.
